README.md
===========
PynaColaDa
===========


Software package for easy and customized Processing huge amounts of gridded Climate Data.

Imagine that your tons of terrabytes of data can be processed in space or time
(or combination) with only a few readable(!) lines of code. On the one hand, we
would obviously fall back to tools like cdo or ncl for such tasks. However, we
are stuck to the limited data analysis options that these provide. On the other
hand, scientific software packages like Matlab, R and IDL (and also Python of
course!) provide very customizable climate data analysis tools, yet they have
severe memory restrictions due to the limited RAM in computers. They also don't
allow a quick moviewize visualization of such large data sets, simply because
they cannot load all those 'terrabytes' into memory at once. The
PynaColaDa-tool now provides the best of both worlds: it allows to perform any
arbitrary pre-defined or user-defined custom function/analysis on a massively
HUGE dataset very easily, with great performance (interfacing with the numpy
library), and WITHOUT the memory restrictions! As it both directly 'reads from'
and 'writes to' NetCDF format, it allows to easily scroll through your data
analysis, i.e. with ncview.

This package also contains some convenient netcdf dataformat
extraction  (see ncgettypecode, ncgetdatetime ) and construction tools
(see ncwritedatetime, nccopydimension, nccopyattrfile, nccopyattrvar,
nccopyvariable,csv2netcdf)

The power of tools lies in their examples. Typical usage of the package, in particular pynacolada.pcd(), can be found below. Additional examples can be found under the 'examples-directory'.::


    import pynacolada as pcd
    import numpy as np
    import datetime as dt
    import sciproc as sp
    from time import sleep
    from Scientific.IO import NetCDF
    
    # purpose: A short tutorial of using pynacolada.pcd().  The example should work with COSMO-CLM data output.
    # author: Hendrik Wouters <hendrik.wouters@ees.kuleuven.be>
    
    print('example 1: calculate the mean temperature of entire vertical columns (per    \
           lat-lon grid cell)')
    fnin ='/media/URB_UIP_2/data/rcm/0.009_20020101_berlin/cclm/out01/lffd2002050100.nc'
    print('reading:',fnin)
    fin = NetCDF.NetCDFFile(fnin,'r')
    datin =  [{'file': fin, \
               'varname': 'T', \
              },\
             ]
    
    fnout = '/media/URB_UIP_2/data/rcm/0.009_20020101_berlin/cclm/anatest0.nc'
    
    fout = NetCDF.NetCDFFile(fnout,'w')
    datout = [{'file': fout, \
               'varname': 'T'}]
    print(' the function definition: take the mean along the axis. Note that we need to \
            preserve the "amount" of dimensions of the output data compared to the input\
            data.')
    
    func = lambda x: np.array([np.mean(x)])
    dnamsel = ['level',]
    pcd.pcd(func,dnamsel,datin,datout)
    
    fout.close()
    fin.close()
    
    print("previous example didn't go very fast! We try again with the 'example 1', in \
           which we use the appenddim-option, and just generating exactly the same     \
           output! It results in only one call of the function, instead of 38025 times \
           in a some python loop. Note that we need to specify the 'axis=0' in the \
           function definition.")
    sleep(5)
    print("example 1: bis... calculate the mean temperature of entire vertical columns \
           (per lat-lon grid cell)")
    fnin ='/media/URB_UIP_2/data/rcm/0.009_20020101_berlin/cclm/out01/lffd2002050100.nc'
    print('reading:',fnin)
    fin = NetCDF.NetCDFFile(fnin,'r')
    datin =  [{'file': fin, \
               'varname': 'T', \
              },\
             ]
    
    fnout = '/media/URB_UIP_2/data/rcm/0.009_20020101_berlin/cclm/anatest1.nc'
    
    fout = NetCDF.NetCDFFile(fnout,'w')
    datout = [{'file': fout, \
               'varname': 'T'}]
    
    func = lambda x: np.array([np.mean(x,axis=0 )])
    dnamsel = ['level',]
    # here is the appenddim option -> it results in only one call of the function,     \
      instead of 38025- times!
    pcd.pcd(func,dnamsel,datin,datout,appenddim=True)
    
    fout.close()
    fin.close()
    sleep(5)
    print("That was much faster isn't it?")
    
    print('example 2: calculate the mean temperature of the first 10 layers (per    \
           lat-lon grid cell)')
    fnin ='/media/URB_UIP_2/data/rcm/0.009_20020101_berlin/cclm/out01/lffd2002050100.nc'
    print('reading:',fnin)
    fin = NetCDF.NetCDFFile(fnin,'r')
    datin =  [{'file': fin, \
               'varname': 'T', \
    # we specify a 'subspace' using the 'dsel' option
               'dsel': {'level' : range(30,40,1)}, \
              },\
             ]
    
    fnout = '/media/URB_UIP_2/data/rcm/0.009_20020101_berlin/cclm/anatest2.nc'
    
    fout = NetCDF.NetCDFFile(fnout,'w')
    datout = [{'file': fout, \
               'varname': 'T'}]
    func = lambda x: np.array([np.mean(x,axis=0 )])
    dnamsel = ['level',]
    pcd.pcd(func,dnamsel,datin,datout,appenddim=True)
    
    fout.close()
    fin.close()
    
    print(' example 3: calculate the scalar horizontal mean wind speed of the first \
            10 layers ')
    fnin ='/media/URB_UIP_2/data/rcm/0.009_20020101_berlin/cclm/out01/lffd2002050100.nc'
    print('reading:',fnin)
    fin = NetCDF.NetCDFFile(fnin,'r')
    datin =  [{'file': fin, \
               'varname': 'U', \
               'dsel': {'level' : range(30,40,1)}, \
    # The U and V wind speed components are not exactly on the same grid so have other
    # (but similar) coordinates ! Therefore, we need to define dimension aliases (thus
    # preventing pcd from generating a huge dataset)!
               'daliases': { 'srlat':'rlat', 'srlon':'rlon' },\
              },\
    # a second variable input definition
              {'file': fin, \
               'varname':'V', \
               'dsel': {'level' : range(30,40,1)},
    # Same here as for the U component....
               'daliases': { 'srlat':'rlat', 'srlon':'rlon' },\
               }\
             ]
    
    fnout = '/media/URB_UIP_2/data/rcm/0.009_20020101_berlin/cclm/anatest3.nc'
    
    fout = NetCDF.NetCDFFile(fnout,'w')
    datout = [{'file': fout, \
               'varname': 'u'}]
    func = lambda U,V: np.array([np.mean(np.sqrt(U**2+V**2),axis=0 )])
    dnamsel = ['level',]
    pcd.pcd(func,dnamsel,datin,datout,appenddim=True)
    # Note also that the calculation is not entirely correct because the U and V
    # locations are shifted by 1/2-gridcell!
    
    fout.close()
    fin.close()
    
    
    print("example 4: now we get to it's real power. Do the same as before,       \
           but for an hourly time series. ")
    
    # the timeseries: just a list of all the files in the right order
    dts = sp.dtrange(dt.datetime(2002,5,1),dt.datetime(2002,7,1),dt.timedelta(hours=1))
    path = '/media/URB_UIP_2/data/rcm/0.009_20020101_berlin/cclm/out01/'
    dtfiles = [path+'lffd'+dt.datetime.strftime(e,"%Y%m%d%H")+'.nc' for e in dts]
    
    datin =  [{'file': dtfiles, \
               'varname': 'U', \
               'dsel': {'level' : range(30,40,1)}, \
               'daliases': { 'srlat':'rlat', 'srlon':'rlon' },\
              },\
              {'file': dtfiles, \
               'varname':'V', \
               'dsel': {'level' : range(30,40,1)},
               'daliases': { 'srlat':'rlat', 'srlon':'rlon' },\
               }\
             ]
    
    fnout = '/media/URB_UIP_2/data/rcm/0.009_20020101_berlin/cclm/anatest4.nc'
    
    fout = NetCDF.NetCDFFile(fnout,'w')
    datout = [{'file': fout, \
               'varname': 'u'}]
    
    # we write the datetimes to the netcdf file (we could do it also afterwards 
    # after the pcd.pcd()-call)
    pcd.ncwritedatetime(fout,dts,tunits='days', refdat=dt.datetime(2002,1,1))
    
    # function definition:
    func = lambda U,V: np.array([np.mean(np.sqrt(U**2+V**2),axis=0 )])
    dnamsel = ['level',]
    pcd.pcd(func,dnamsel,datin,datout,appenddim=True)
    
    fout.close(); print 'output data written to:',fnout
    print('and enjoy watching your processed data with ncview!')











A Section
=========

No manual Yet.




CHANGES.txt



VERSION 0.2.21: sub revision - di dec 17 15:36:58 CET 2013

changes in csv2netcdf: remove special characters in labels
add example csv2netcdf


VERSION 0.2.20: sub revision - ma dec  2 13:17:02 CET 2013

bugfix in the detemination of the array-position


VERSION 0.2.19: sub revision - za nov 30 22:12:39 CET 2013

adding a more advanced 'stein-alpert' example
some serious bugfixes concerning 'dsel'. Now, it is always defined from the beginning, even when it is not specified by the user ->in that case, data[i]['dsel'] it is equal to range(data[i]['dims']). 

VERSION 0.2.18: sub revision - wo nov 27 14:54:38 CET 2013

whoops, previous versions above 0.2.14 won't work


VERSION 0.2.14: sub revision - wo nov 27 10:02:56 CET 2013

add a forcearray option so that the function number of input dimensions is at least 1.


VERSION 0.2.13: sub revision - di nov 26 14:24:12 CET 2013

bugfix in the construction of the function output dimensions


VERSION 0.2.12: sub revision - zo nov 24 19:18:50 CET 2013

Add README.md to MANIFEST.in, so that it can be installed with pypi-install


VERSION 0.2.11: sub revision - zo nov 17 20:49:35 CET 2013

ignore the copyvalues option when a variable doesn't contain any values.


VERSION 0.2.10: sub revision - zo nov 17 15:06:04 CET 2013

add example for copying values
In order to support without memory problems very large NetCDF variables when copying them, the 'copyvalues' option in 'nccopyvariable' now uses pynacolada.pcd() to copy the values.
bugfix for the case when the output netcdf variable already exists


VERSION 0.2.9: sub revision - zo nov 17 09:08:39 CET 2013

clean up code to get rid of unnecessary dependencies
remove ncgetformat function. It has the same functionality as nctypecode. The latter is renamed to ncgettypecode
Some additions in the README.md
Add documentation to all the procedures
Additional example: pcdsellonlatbox.py


VERSION 0.2.8: sub revision - za nov 16 10:12:37 CET 2013

clean up type() calls
update examples: calculating the running time-average (examples/pcdfuncionexamples.py)
syntax code clean-up (pynacolada/pynacolada.py)
predim is now defined as a data input attribute

VERSION 0.2.7: sub revision - vr nov 15 12:18:37 CET 2013

load function modules from ncdfextract.py (forgotten 'return' at the end of __init.py__)


VERSION 0.2.6: sub revision - wo nov 13 15:32:40 CET 2013

bug fixes: 
- As expected by 'readicecubeps', the dimension that represents the 'list' of files should always be the first (outer) dimension!
- Create arrays with dimension-sizes corresponding to the lengths of dimnoiter-items, and not to the dimnoiter-items itself.


VERSION 0.2.4: sub revision - di nov 12 10:26:08 CET 2013

rename pcl to pcd to be more consistent to the package name and its purpose.


VERSION 0.2.3: sub revision - ma nov 11 22:19:44 CET 2013

update description of the package in README.md


VERSION 0.2.2: sub revision - ma nov 11 21:31:46 CET 2013

some readability changes in the examples.


VERSION 0.2.1: sub revision - zo nov 10 21:47:04 CET 2013

make it python3 compatible (hopefully) by providing brackets for the 'print' statements
remove unnecessary print statements
add a more extensive example with netcdf data from COSMO-CLM (examples/testana.py). This is also added to the README.md


VERSION 0.2.0: minor revision - zo nov 10 17:20:52 CET 2013

adding additional handy procedures for doing basic NetCDF operations from deprecated ncdftools package so that pynacolada becomes independent of it.
moving README.txt to README.md (github standard)


VERSION 0.1.0: minor revision - zo nov 10 15:12:39 CET 2013

Adding information on how to install package (see INSTALL)
Adding build-files to package (cfr. ./ubuntubuild.sh)
bugfix when defining the extra dimension for a file list (still needs to be tested)
adding some

VERSION 0.0.3: sub revision - vr nov  8 15:00:04 CET 2013

Add exmple directory and licence


VERSION 0.0.2: sub revision - vr nov  8 14:55:00 CET 2013


example update in README.txt

VERSION 0.0.1: sub revision - vr nov  8 14:38:12 CET 2013


Initial version. 

Current options that implemented:

 - implement dimension aliases (-> implemented as 'daliases'-option for the data input definition)
 - allow multiple input files in a stream (such as hourly climate data). Therefore, one needs to define input data as a filename-list
 - select custom coordinate frames (-> implemented as 'dsel'-option for data input definition)
 - detect output dtype, first from output data file, then from output function
 - sensible determination of memory limitation (-> maxmembytes)
 - allow wild cards in specifying function dimensions (-> appenddim!)


examples/pynatest.py
#!/usr/bin/env python

# Example 1: calculate the mean scalar wind speed of the first 10 layers
import pynacolada as pcd
from Scientific.IO import NetCDF
import numpy as np

fnin = '/home/hendrik/data/belgium_aq/rcm/aq09/stage1/int2lm/laf2009010100_urb_ahf.nc'
#print fnin
# fobjin = open(fnin,'rb')
fin = NetCDF.NetCDFFile(fnin,'r')
fnout = 'laf2009010100_out.nc'
#print fnout
# fobjout = open(fnout,'wb+''rlat')
fout = NetCDF.NetCDFFile(fnout,'w')
# input data definitions
datin =  [{'file': fin, \
           'varname': 'U', \
           'dsel': {'level' : range(30,40,1)}, \
           'daliases': { 'srlat':'rlat', 'srlon':'rlon' },\
          },\
          {'file': fin, \
           'varname':'V', \
           'dsel': {'level' : range(30,40,1)},
           'daliases': { 'srlat':'rlat', 'srlon':'rlon' },\
           }\
         ]
# output data definitions
datout = [{'file': fout, \
           'varname': 'u'}]
# function definition:
func = lambda U,V: np.array([np.mean(np.sqrt(U**2+V**2),axis=0 )])
dnamsel = ['level',]
pcd.pcd(func,dnamsel,datin,datout,appenddim=True)
print 'output data written to:',fnout
